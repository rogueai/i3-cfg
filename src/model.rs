use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Default, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Config {
    pub keybindings: Vec<Keybinding>,
}

#[derive(Serialize, Deserialize, Default, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Keybinding {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub group: Option<String>,
    pub modifiers: Vec<String>,
    #[serde(flatten)]
    pub variant: KeybindingVariant,
    pub command: String,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(untagged, rename_all = "camelCase")]
pub enum KeybindingVariant {
    Keysym { key: String },
    Keycode { code: u8 },
    Unknown,
}

impl Default for KeybindingVariant {
    fn default() -> Self {
        KeybindingVariant::Unknown
    }
}
