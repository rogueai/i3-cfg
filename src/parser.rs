use crate::model::{Config, Keybinding, KeybindingType};
use pest::iterators::Pair;
use pest::Parser;

#[derive(Parser)]
#[grammar = "grammar/i3-cfg.pest"]
pub struct I3Parser;

pub fn parse(config: String) -> Config {
    let config = I3Parser::parse(Rule::config, &config)
        .expect("parse failed") // unwrap the parse result
        .next()
        .unwrap(); // get and unwrap the `file` rule; never fails

    let mut keybindings: Vec<Keybinding> = vec![];
    parse_pair(config, &mut keybindings);

    Config { keybindings }
}
/// Recursively parse [Pair]s and populate keybindings.
fn parse_pair(parent: Pair<Rule>, keybindings: &mut Vec<Keybinding>) {
    for child in parent.into_inner() {
        match child.as_rule() {
            Rule::stmt
            | Rule::expr
            | Rule::keybinding
            | Rule::bindsym
            | Rule::bindcode
            | Rule::bindmouse => {
                parse_pair(child, keybindings);
            }
            Rule::binding_sym | Rule::binding_code | Rule::binding_mouse => {
                keybindings.push(Keybinding::default());
                parse_pair(child, keybindings);
            }
            Rule::keycode => {
                let mut kb = keybindings.pop().unwrap();
                kb.variant = KeybindingType::Keycode;
                kb.key = child.as_str().parse().unwrap();
                keybindings.push(kb);
            }
            Rule::keysym => {
                let mut kb = keybindings.pop().unwrap();
                kb.variant = KeybindingType::Keysym;
                kb.key = child.as_str().parse().unwrap();
                keybindings.push(kb);
            }
            Rule::button => {
                let mut kb = keybindings.pop().unwrap();
                kb.variant = KeybindingType::Button;
                kb.key = child.as_str().parse().unwrap();
                keybindings.push(kb);
            }
            Rule::group => {
                let mut kb = keybindings.pop().unwrap();
                kb.group = Some(String::from(child.as_str()));
                keybindings.push(kb);
            }
            Rule::modifier => {
                let mut kb = keybindings.pop().unwrap();
                kb.modifiers.push(String::from(child.as_str()));
                keybindings.push(kb);
            }
            Rule::criteria => {
                let mut kb = keybindings.pop().unwrap();
                kb.criteria = Some(String::from(child.as_str()));
                keybindings.push(kb);
            }
            Rule::command => {
                let mut kb = keybindings.pop().unwrap();
                kb.command = String::from(child.as_str());
                keybindings.push(kb);
            }
            Rule::EOI => (),
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::ser;

    #[test]
    fn test_keysym() {
        let config = r###"
        bindsym Mod4+u border none
        "###;
        let config = parse(String::from(config));
        let json = serde_json::to_string_pretty(&config).unwrap();
        assert_eq!(
            json,
            r###"
        {
            "config"
        }
        "###
        )
    }
}
