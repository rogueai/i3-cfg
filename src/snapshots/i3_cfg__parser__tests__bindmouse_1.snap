---
source: src/parser.rs
expression: config
---
{
  "keybindings": [
    {
      "type": "button",
      "modifiers": [],
      "key": "2",
      "command": "kill"
    }
  ]
}
